### PR TITLE
Update Installer to support main package install and DDLM

### PR DESCRIPTION
**Pull request by dkbennett**
Implementation for #607 

Adds support for installing main packages via the installer. This is intended to address the scenario where Project Reunion consists of framework packages and a Main package, along with a DDLM package, which is a main package.
* Adds install logic to handle differentiating frameworks from main packages and matching main packages with system native architecture.
* Creates more test packages for Project Reunion Main packages and DDLM packages
* Adds test packages to the resource list for inclusion into the installer.
* Updates tests to verify correct deployment of new main packages.

Verified by running updated tests successfully and with some manual testing of edge cases around provisioning.

### CODE REVIEW

File: installer/ProjectReunionInstall.sln [hunk 1]
```diff
@@ -24,6 +29,16 @@ Global
 		Release|x86 = Release|x86
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{E6E59B30-9F55-4550-AA73-3B3B3DC89872}.Debug_test|Any CPU.ActiveCfg = Release|x64
+		{E6E59B30-9F55-4550-AA73-3B3B3DC89872}.Debug_test|Any CPU.Build.0 = Release|x64
+		{E6E59B30-9F55-4550-AA73-3B3B3DC89872}.Debug_test|ARM.ActiveCfg = Release|x64
```
Comment (DrusTheAxe) [inline installer/ProjectReunionInstall.sln:34]: What's the point of Debug_test? Debug+Release I get but why Debug_test rather than just Debug?

File: installer/dev/ProjectReunionInstall.rc [hunk 2]
```diff
@@ -17,6 +17,26 @@ PR_FRAMEWORK_X64_RCENTRY
 PR_FRAMEWORK_ARM64_RCENTRY
 #endif
 
+#if defined(PR_MAIN_X86_RCENTRY)
```
Comment (DrusTheAxe) [inline installer/dev/ProjectReunionInstall.rc:20]: PR's a bit cryptic (too many possibilities).

SUGGESTION: `PROJECTREUNION_...`

Besides when(!) the coming name change arrives that'll be easy to track down all the places to change

File: installer/dev/install.cpp [hunk 3]
```diff
@@ -10,51 +10,82 @@ using namespace Windows::System;
...
 
-    winrt::hresult AddPackage(const Uri& packageUri)
+    HRESULT RegisterPackage(const std::wstring& packageFullName)
+    {
+        PackageManager packageManager;
+        const auto deploymentOperation{ packageManager.RegisterPackageByFullNameAsync(packageFullName, nullptr, DeploymentOptions::None) };
+        deploymentOperation.get();
+        if (deploymentOperation.Status() != AsyncStatus::Completed)
+        {
+            return static_cast<HRESULT>(deploymentOperation.ErrorCode());
```
Comment (DrusTheAxe) [inline installer/dev/install.cpp:20]: ErrorCode or ExtendedErrorCode?

Comment (DrusTheAxe) [inline installer/dev/install.cpp:20]: Register isn't expected to fail so `return LOG_IF_FAILED(static_cast<HRESULT>(...));` would be useful

File: installer/dev/install.cpp [hunk 4]
```diff
@@ -10,51 +10,82 @@ using namespace Windows::System;
...
         const auto deploymentOperation{ packageManager.AddPackageAsync(packageUri, nullptr, DeploymentOptions::None) };
         deploymentOperation.get();
         if (deploymentOperation.Status() != AsyncStatus::Completed)
         {
-            return deploymentOperation.ErrorCode();
+            auto hrAddPackage = static_cast<HRESULT>(deploymentOperation.ErrorCode());
+            if (hrAddPackage == ERROR_PACKAGE_ALREADY_EXISTS)
+            {
+                // Package already exists (such as via provisioning), re-register it instead.
+                return RegisterPackage(packageProperties->fullName.get());
```
Comment (DrusTheAxe) [inline installer/dev/install.cpp:37]: RETURN_IF_FAILED?

File: installer/dev/install.cpp [hunk 5]
```diff
@@ -10,51 +10,82 @@ using namespace Windows::System;
...
+            return true;
+        }
+
         SYSTEM_INFO systemInfo{};
         GetNativeSystemInfo(&systemInfo);
         const auto systemArchitecture{ static_cast<ProcessorArchitecture>(systemInfo.wProcessorArchitecture) };
 
-        // Neutral package architecture is applicable on all systems.
-        if (arch == ProcessorArchitecture::Neutral)
+        // Same-arch is always applicable for any package type.
```
Comment (DrusTheAxe) [inline installer/dev/install.cpp:72]: NIT: architecture

File: installer/dev/install.cpp [hunk 6]
```diff
@@ -10,51 +10,82 @@ using namespace Windows::System;
...
+        // so only the same-architecture or neutral will be matched for non-frameworks.
+        if (!packageProperties->isFramework)
         {
-            return true;
+            return false;
         }
 
+        // Framework packages have additional logic.
+
         // On x64 systems, x86 architecture is also applicable.
```
Comment (DrusTheAxe) [inline installer/dev/install.cpp:87]: Isn't the ARM64 case similar?

File: installer/dev/install.cpp [hunk 7]
```diff
@@ -108,6 +139,13 @@ namespace ProjectReunionInstaller {
         properties->architecture = static_cast<ProcessorArchitecture>(arch);
         THROW_IF_FAILED(id->GetVersion(&properties->version));
 
+        // Populate framework from the manifest properties.
+        wil::com_ptr<IAppxManifestProperties> manifestProperties;
+        THROW_IF_FAILED(manifest->GetProperties(wil::out_param(manifestProperties)));
+        BOOL isFramework = FALSE;
```
Comment (DrusTheAxe) [inline installer/dev/install.cpp:145]: NIT: `BOOL isFramework{};`

File: installer/dev/install.cpp [hunk 8]
```diff
@@ -108,6 +139,13 @@ namespace ProjectReunionInstaller {
...
+        THROW_IF_FAILED(manifestProperties->GetBoolValue(L"Framework", &isFramework));
+        properties->isFramework = isFramework == TRUE;
```
Comment (DrusTheAxe) [inline installer/dev/install.cpp:147]: NIT: please use separate assignments e.g.

```
isFramework = TRUE;
properties->isFramework = TRUE;
```

or

```
isFramework = TRUE;
properties->isFramework = isFramework;
```

File: installer/dev/install.cpp [hunk 9]
```diff
@@ -156,20 +194,24 @@ namespace ProjectReunionInstaller {
 
         // Add the package
         Uri packageUri{ packageFilename };
-        hresult hrAddResult = AddPackage(packageUri);
+        auto hrAddResult = AddPackage(packageUri, packageProperties);
```
Comment (DrusTheAxe) [inline installer/dev/install.cpp:197]: NIT: `const auto hrAddResult{ AddPackage(...) };`

File: installer/dev/install.cpp [hunk 10]
```diff
@@ -156,20 +194,24 @@ namespace ProjectReunionInstaller {
...
         if (!quiet)
         {
-            std::wcout << "Package deployment result : 0x" << std::hex << hrAddResult.value << std::endl;
+            std::wcout << "Package deployment result : 0x" << std::hex << hrAddResult << std::endl;
```
Comment (DrusTheAxe) [inline installer/dev/install.cpp:200]: Blech. printf is so much more dense and easier to read. Alas, no [std::format](https://en.cppreference.com/w/cpp/utility/format/format) until C++20. But wprintf will be easier than cout to transition to std::format so I'd recommend it even today

File: installer/dev/install.h [hunk 11]
```diff
@@ -10,6 +10,7 @@ namespace ProjectReunionInstaller {
         wil::unique_cotaskmem_string familyName;
         winrt::Windows::System::ProcessorArchitecture architecture{ winrt::Windows::System::ProcessorArchitecture::Unknown };
         std::uint64_t version{};
+        bool isFramework = false;
```
Comment (DrusTheAxe) [inline installer/dev/install.h:13]: NIT: `bool isFramework{};`

File: installer/dev/project_reunion_definitions.h [hunk 12]
```diff
@@ -39,15 +39,56 @@
 #define PR_FRAMEWORK_ARM64_RCTYPE PACKAGE
 #define PR_FRAMEWORK_ARM64_PATH "..\\test\\testpackages\\framework_arm64.msix"
 
+// x86 Main
+#define PR_MAIN_X86_ID L"PR_MAINPACKAGE_X86"
```
Comment (DrusTheAxe) [inline installer/dev/project_reunion_definitions.h:43]: NIT: MixedCase would be easier to read than ALLUPPERCASE e.g. `ProjectReunion_MainPackage_x86` and `ProjectReunion_DdlmPackage_x86`. Or is there a reason I'm missing for uppercase?

File: installer/dev/project_reunion_definitions.h [hunk 13]
```diff
@@ -57,6 +98,30 @@
...
+#define PR_MAIN_X86_LISTENTRY { PR_MAIN_X86_ID, PR_MAIN_X86_TYPE },
+#define PR_MAIN_X86_RCENTRY PR_MAIN_X86_RCID PR_MAIN_X86_RCTYPE PR_MAIN_X86_PATH
+
+#define PR_MAIN_X64_LISTENTRY  { PR_MAIN_X64_ID, PR_MAIN_X64_TYPE },
+#define PR_MAIN_X64_RCENTRY PR_MAIN_X64_RCID PR_MAIN_X64_RCTYPE PR_MAIN_X64_PATH
+
+// arm64 is included only on arm64 builds
+#if defined(_M_ARM64)
+    #define PR_MAIN_ARM64_LISTENTRY { PR_MAIN_ARM64_ID, PR_MAIN_ARM64_TYPE },
+    #define PR_MAIN_ARM64_RCENTRY PR_MAIN_ARM64_RC     PR_MAIN_ARM64_RCTYPE                 PR_MAIN_ARM64_PATH
```
Comment (DrusTheAxe) [inline installer/dev/project_reunion_definitions.h:110]: Weird spacing. Accidental indentation? Tabs? Leftover indentation from an earlier simpler day when it was nicely lined up w/o any conditional logic or like complexities?

Same comment applies throughout

File: installer/test/CreateInstallerTestPackages/appxmanifest_ddlm_x64.xml [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10"
+  xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10"
+  xmlns:uap3="http://schemas.microsoft.com/appx/manifest/uap/windows10/3"
+  xmlns:uap5="http://schemas.microsoft.com/appx/manifest/uap/windows10/5"
+  xmlns:com="http://schemas.microsoft.com/appx/manifest/com/windows10"
+  xmlns:rescap="http://schemas.microsoft.com/appx/manifest/foundation/windows10/restrictedcapabilities"
+  IgnorableNamespaces="uap uap3 uap5 com rescap">
+
+  <Identity
+    Name="ProjectReunion.Test.DDLM-x64"
```
Comment (DrusTheAxe) [inline installer/test/CreateInstallerTestPackages/appxmanifest_ddlm_x64.xml:13]: SUGGESTION: Use more specific package names to distinguish your test packages from others to avoid accidental collision (especially when investigating test failures), e.g. `Name="ProjectReunion.Installer.Test.DDLM-x64"`

Same comment applies throughout

File: installer/test/InstallerFunctionalTests/FunctionalTests.cpp [hunk 15]
```diff
@@ -25,49 +25,60 @@ namespace ProjectReunionInstallerTests
         TEST_METHOD(VerifyInstallerRuns)
         {
             auto result = RunInstaller();
-            Assert::AreEqual(result, S_OK);
+            Assert::AreEqual(S_OK, result);
```
Comment (DrusTheAxe) [inline installer/test/InstallerFunctionalTests/FunctionalTests.cpp:28]: VSTest?

RECOMMNEDATION: Use TAEF.

DynamicDependencies' tests use VSTest and given that 1st hand experience I'm in the process of rewriting them to TAEF... per [Rewrite DynamicDependencies tests using TAEF #559](https://github.com/microsoft/ProjectReunion/issues/559)

If not in this PR is there an Issue for follow up?

File: installer/ProjectReunionInstall.sln (hunk 1 above)
Comment (dkbennett) [reply installer/ProjectReunionInstall.sln:34]: I don't know, I didn't hand-edit that, it's something VS populated. Should it be removed/altered? How so?

File: installer/dev/ProjectReunionInstall.rc (hunk 2 above)
Comment (dkbennett) [reply installer/dev/ProjectReunionInstall.rc:20]: I ran into character limits on the macro expression for the resource compiler. I concur that spelling out the name would be better but I'm adjusting to limitations in the resource compiler and feel this is the best option given those limitations.

File: installer/test/InstallerFunctionalTests/FunctionalTests.cpp (hunk 15 above)
Comment (dkbennett) [reply installer/test/InstallerFunctionalTests/FunctionalTests.cpp:28]: I had difficulty using TAEF earlier, added an issue for followup, #714

File: installer/dev/project_reunion_definitions.h (hunk 12 above)
Comment (dkbennett) [reply installer/dev/project_reunion_definitions.h:43]: I followed existing pattern in resource scripts.

File: installer/dev/install.cpp (hunk 3 above)
Comment (dkbennett) [reply installer/dev/install.cpp:20]: ErrorCode, as per the documented sample for the Deployment API. 

https://docs.microsoft.com/en-us/uwp/api/windows.management.deployment.packagemanager.addpackageasync?view=winrt-19041

Comment (dkbennett) [reply installer/dev/install.cpp:20]: In my manual testing I recall it did fail, and the documentation does not indicate that it is not expected to fail, though there is no sample for register.

File: installer/dev/install.cpp (hunk 4 above)
Comment (dkbennett) [reply installer/dev/install.cpp:37]: Fixed

File: installer/dev/project_reunion_definitions.h (hunk 13 above)
Comment (dkbennett) [reply installer/dev/project_reunion_definitions.h:110]: Yes leftover, fixed.

File: installer/test/CreateInstallerTestPackages/appxmanifest_ddlm_x64.xml (hunk 14 above)
Comment (dkbennett) [reply installer/test/CreateInstallerTestPackages/appxmanifest_ddlm_x64.xml:13]: I distinguished names slightly differently but did add in the "Installer" keyword to make them more distinct and avoid accidental collisions and easier troubleshooting.

File: installer/ProjectReunionInstall.sln (hunk 1 above)
Comment (DrusTheAxe) [reply installer/ProjectReunionInstall.sln:34]: VS automagically adding it due to something something vcxproj something?

Do you have any C# in your solution?

@BenJKuhn @kennykerr  @Scottj1s @aeloros can you explain this?

File: installer/dev/install.cpp (hunk 3 above)
Comment (DrusTheAxe) [reply installer/dev/install.cpp:20]: TL;DR the sample's old and not best practice. You want to return ExtendedErrorCode.

Yes, but that sample's very old, dating back to Win8. It returns the generic error code per PackageManager.AddPackageAsync() e.g. the dreaded 0x80073CF9 aka HRESULT_FROM_WIN32(ERROR_INSTALL_FAILED).

[.ExtendedErrorCode](https://docs.microsoft.com/uwp/api/windows.management.deployment.deploymentresult.extendederrorcode?view=winrt-19041) was added shortl;y after to expose more meaningful error codes -- AccessDenied, OutOfDiskSpace, and others.

When errors happen the actual error code boils up to the API but it's almost always remapped into the general 80073CF9 "Install Failed". You have to hunt down Deployment's log to determine the actual failure. Or call `.ExtendedErrorCode`. This was intentional and well intentioned in Win8 but, 20-20 hindsight, was a mistake. Alas, PackageManager can't directly fix this without compatibility issues, hence the added property.

If the installer returns `deploymentOperation.ErrorCode` all folks will know is "Install Failed". If the installer returns `deploymentOperation.ExtendedErrorCode` the actual (rich) error code will be readily known.

We should also file a sample bug to update it given current best practices.

**Review by DrusTheAxe — APPROVED** [review]
